pkg/apis/apis/v1alpha1/types_apiresourceschema.go

```diff
@@ -163,3 +163,103 @@ func (v *APIResourceVersion) SetSchema(schema *apiextensionsv1.JSONSchemaProps)
+	// version that would otherwise be lost during round-tripping to a different API version. An example key and value
+	// might look like this: v2.conversion.apis.kcp.io/preserve: {"spec.someNewField": "someValue"}.
+	VersionPreservationAnnotationKeySuffix = ".conversion.apis.kcp.io/preserve"
+)
+
+// +crd
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:resource:scope=Cluster,categories=kcp
+// +kubebuilder:printcolumn:name="Age",type="date",JSONPath=".metadata.creationTimestamp"
+
+// APIConversion contains rules to convert between different API versions in an APIResourceSchema. The name must match
+// the name of the APIResourceSchema for the conversions to take effect.
+type APIConversion struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
+
+	// Spec holds the desired state.
+	Spec APIConversionSpec `json:"spec"`
+}
+
+// APIConversionSpec contains rules to convert between different API versions in an APIResourceSchema.
+type APIConversionSpec struct {
+	// +required
+	// +listType=map
+	// +listMapKey=from
+	// +listMapKey=to
+
+	// conversions specify rules to convert between different API versions in an APIResourceSchema.
+	Conversions []APIVersionConversion `json:"conversions"`
+}
+
+// APIVersionConversion contains rules to convert between two specific API versions in an
+// APIResourceSchema. Additionally, to avoid data loss when round-tripping from a version that
+// contains a new field to one that doesn't and back again, you can specify a list of fields to
+// preserve (these are stored in annotations).
+type APIVersionConversion struct {
+	// +required
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:Pattern=^v[1-9][0-9]*([a-z]+[1-9][0-9]*)?$
+
+	// from is the source version.
+	From string `json:"from"`
+
+	// +required
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:Pattern=^v[1-9][0-9]*([a-z]+[1-9][0-9]*)?$
+
+	// to is the target version.
+	To string `json:"to"`
+
+	// +required
+	// +listType=map
+	// +listMapKey=destination
+
+	// rules contains field-specific conversion expressions.
+	Rules []APIConversionRule `json:"rules"`
+
+	// +optional
+
+	// preserve contains a list of JSONPath expressions to fields to preserve in the originating version
+	// of the object, relative to its root, such as '.spec.name.first'.
+	Preserve []string `json:"preserve,omitempty"`
+}
+
+// APIConversionRule specifies how to convert a single field.
+type APIConversionRule struct {
+	// +required
+	// +kubebuilder:validation:MinLength=1
+
+	// field is a JSONPath expression to the field in the originating version of the object, relative to its root, such
+	// as '.spec.name.first'.
+	Field string `json:"field"`
+
+	// +required
+
+	// destination is a JSONPath expression to the field in the target version of the object, relative to
+	// its root, such as '.spec.name.first'.
```

Comment:
nit: it might be useful to have one coherent example that all the GoDoc references, for clarity